compiler/src/dotty/tools/dotc/ast/DesugarEnums.scala

```diff
@@ -84,35 +92,55 @@ object DesugarEnums {
   private def valuesDot(name: PreName)(implicit src: SourceFile) =
     Select(Ident(nme.DOLLAR_VALUES), name.toTermName)
 
-  private def registerCall(using Context): Tree =
-    Apply(valuesDot("register"), This(EmptyTypeIdent) :: Nil)
+  private def ArrayLiteral(values: List[Tree], tpt: Tree)(using Context): Tree =
+    val clazzOf = TypeApply(ref(defn.Predef_classOf.termRef), tpt :: Nil)
+    val ctag    = Apply(TypeApply(ref(defn.ClassTagModule_apply.termRef), tpt :: Nil), clazzOf :: Nil)
+    val apply   = Select(ref(defn.ArrayModule.termRef), nme.apply)
+    Apply(Apply(TypeApply(apply, tpt :: Nil), values), ctag :: Nil)
 
-  /**  The following lists of definitions for an enum type E:
+  /**  The following lists of definitions for an enum type E and known value cases e_0, ..., e_n:
    *
-   *   private val $values = new EnumValues[E]
-   *   def values = $values.values.toArray
+   *   private val $values = Array[E](e_0,...,e_n)(ClassTag[E](classOf[E]))
+   *   @annotation.threadUnsafe private lazy val $valuesReverse =
```

Comment:
This should most likely be thread-safe, shouldn't it?

Reply:
I thought not because `$values` is loaded first in `<cinit>` which is thread safe and then this is just a view

Reply:
But two threads can call `valueOf` at the same, resulting in several threads trying to access `$valuesReverse` at the same time.

Reply:
is this worse than two threads calling a `def` at the same time? - Im going to get rid of it anyway if there are no objections to the "bytecode bloat" for the switch method